### DISCUSSION TITLE
Clustering Of Routes

### DISCUSSION THREAD

**Discussion by saravanapriya266** [google/or-tools#3414] [Routing (and legacy CP) questions]
I am using or-tools VRP to find optimal route with number of locations a truck can go as constraint. I tried with PARALLEL_CHEAPEST_INSERTION and GUIDED_LOCAL_SEARCH as search strategy which gave me minimal vehicle utilisation as we expected but routes are not clustered together. We want same vehicle to visit locations near by. 

To get better clustering of routes when I tried GLOBAL_CHEAPEST_ARC and GUIDED_LOCAL_SEARCH. Clusters are  grouped better with this but number of vehicles got increased. 
**Which strategy would give me lesser vehicle utilisation and lesser travel distance with good clustering of locations**?
![Screenshot from 2022-08-03 06-54-01](https://user-images.githubusercontent.com/16831982/182532804-560562a1-4652-4f12-b160-70c5068cac2a.png)

Example - Route 11 has nodes which are not clustered.

**What version of OR-Tools and what language are you using?**
Version: master/v7.3/v7.2 etc.
Language: Python

**Which solver are you using ? Routing Solver** VRP
**What operating system (Linux, Windows, ...) and version?** Linux

**Comment by jmarca** [comment]
This question doesn't make much sense to me.  I expect that you are not doing anything complicated, and that your routing solver problem is set up to minimize total distance traveled.  

If that is true (hint: supply a working example), then clustering is *not* something that the solver is trying to achieve.  It is instead attempting to minimize total travel distance (or perhaps time).

Clustering is a good strategy when setting up the problem manually, but not necessarily something that will drop out of a solver run.

If you really must have clustering of nodes, you can limit the maximum distance between any two nodes.  In your distance callback, if the distance is larger than the maximum span of your desired cluster, then return a very large number (max int, or 100000000 or something).
That will prevent node to node links that are overly long, and will tend to promote local grouping.  

But of course, you might also get a long snake-type route where each link is at the maximum limit and it continues in one direction.

There are other API calls you might be able to twist into giving you a clustered result, like the SetSpanCostCoefficientForAllVehicles along with a suitably clever dimension definition. 

Again, clustering is *not* a goal of the solver.  Minimizing total distance (or time) is.
